RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,206 @@
+Group 8 solution!
+"""
+
+import logging
+import random
+
+'''Given: There is a parking lot with N spaces and N-1 cars in it.
+The parking lot is described by an array of numbers, that is a permutation of numbers
+from 0 to N. Cars are with numbers from 1 to N-1, 0 means an empty parking space.
+
+E.G. Initial order = [1, 2, 0, 3]
+       Final order = [3, 1, 2, 0]
+
+Task: to rearrange the cars in a given way
+and print a series of moves from position x to y
+
+For our example, the series of moves should be e.g.:
+    move 1 to 2
+    move 3 to 0
+'''
+
+class Movement(object):
+
+    """This class stores information about one movement,
+    i.e. a car moving from position *from_index* to position *to_index*.
+    """
+
+    def __init__(self, car, from_index, to_index):
+        self.car = car
+        self.from_index = from_index
+        self.to_index = to_index
+
+    def __str__(self):
+        move_str = "Move car #%s from %s to %s" %(self.car, self.from_index, self.to_index)
+        return move_str
+
+class RearrangeCars(object):
+
+    '''Class to rearrange the cars and print the moves
+       - Uses car_position_dict to optimize searching the position of a given car
+       - Stores additionally the position of empty spot in current car arrangement
+       self.current_car_order.
+       - Moves are either printed to the console or logged
+       - Stores and updates min_num_steps_left that shows min number of moves left
+        in order to avoid comparison current and final order lists that takes O(n)
+    '''
+
+    def __init__(self, init_car_order, final_car_order):
+
+        '''Initializes class instance with initial, final order and
+        stores the position of empty parking space in the current car order
+        Memory is O(n) to store 3 order lists, 1 car_position_dict and a set of
+        not arranged cars
+
+        Storing car_position_dict helps to search indices of car in O(1) instead of O(n)
+        A set of not_arranged_cars helps to now when to stop in O(1) as well as to draw
+        in O(1) unarranged car
+
+        Arguments:
+           init_car_order: a valid non-empty list of cars + 1 \
+           empty_place given in their parking place order
+
+           final_car_order: a valid non-empty list of cars + 1 \
+           empty_place according to the desired order
+
+        Returns:
+            None
+        '''
+
+        self.initial_car_order = init_car_order
+        self.final_car_order = final_car_order
+        self.print_console = True
+        self._variables_to_init_state()
+
+    def print_moves_log_file(self, moves_list):
+
+        '''Prints a list of moves to the log file
+        
+        Arguments:
+            moves: a list of Movement class instances
+        Returns:
+            None
+        '''
+        
+        info_message = 'Number moves is {0} > 1000, logging them..'.format(len(moves_list))
+        print info_message
+        logging.info('Initial car order:')

```

Comment:
I'm wondering about putting this in a one function with newlines, as that's quite a lot of repetitive calls